accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+
+Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.
+
+Source generators should not rely on referencing nor executing framework types they wish to extend.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  For example, API added to help a source generator do its job may not be present when running on .NET Framework or the .NET Core version the compiler is running on.  There is no type-equivalence guaranteed between types loaded in the compiler process and those of the user code that is under analysis.  For example, a source generator cannot instantiate attribute instances in the compiler process from user references to those attributes in source.  Should a source generator need to identify type usage it will need to examine user code references by name or resolve types in the user's references to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library helper code when generating source.  For example, serializer code which needs to examine the shape of user types can be shared between the runtime library and source generator.  Common abstractions can be made over metadata to allow the same code to handle different sources of metadata (source vs reflection).
+
+Source generators should localize any diagnostic messages displayed to the user.  This can be done using localized resource satellite assemblies.  These assemblies should be distributed with the source generator assembly. 
+
+## Q & A
+
+### Why not just deliver source generators in the SDK same as roslyn analyzers?
+
+Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable target frameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.
```

Comment:
Why do you believe that distributing the generators at the SDK level conflicts with the idea that they should be specific to the target framework? The build logic could easily pick the appropriate version of the generator for the target framework that is being built. Hence I'm unsure why you feel these ideas are in conflict.

Reply:
If the build logic is selecting the appropriate version of the generator than it effectively *is the same* as this proposal: we're just leveraging existing build logic for target framework selection by placing in the ref-pack rather than inventing something new.  This question is addressing the alternative of a single-non-target-framework-versioned copy of generators (as is done with analyzers)